File: 알고리즘 고득점 Kit/해시/폰켓몬.cpp
#include <vector>
#include <map>
using namespace std;

int solution(vector<int> nums)
{
    int answer = 0;
    map<int,int> m;
    
    for(int i=0;i<nums.size();i++){
        
        if(m.find(nums[i])!=m.end()){
            m[nums[i]]++;
        }
        
        else{
            m.insert(make_pair(nums[i],1));
        }
    }
                     
    answer = min(nums.size()/2, m.size());
    
    return answer;
}

File: Java/LV1/폰켓몬.java
import java.util.*;

class Solution {
    public int solution(int[] nums) {
        int answer = 0;
        
        HashMap<Integer,Integer> hm = new HashMap<>();
        
        for(int i=0;i<nums.length;i++){
            int cur = nums[i];
            if(hm.containsKey(cur)){
                int value = hm.get(cur);
                value+=1;
                hm.put(cur,value);
            }
            else{
                hm.put(cur,1);
            }
        }
        
        if(nums.length/2 > hm.size()){
            answer = hm.size();
        }
        else{
            answer = nums.length/2;
        }
        return answer;
    }
}

// 1:1
// 2:5
// 3:2
